### PR TITLE
MRG, ENH/FIX: Make cHPI calculation code public

### PR DESCRIPTION
**Pull request by larsoner**
I spent a lot of time revamping our cHPI localization code to match what MaxFilter does. In particular I needed to:

1. Deal with signs properly in correlation check.
2. Check corrs per coil (not jointly) and skip if at least three match.
3. Remove projections before computing whitener (should not include projs).
4. Choose coils based on each one's individual GOF (and later by translated distance).
5. Iteratively subselect of coils from the N good ones.
6. Use a low-order SSS expansion to improve fits by suppressing noise.
7. Use a 1cm grid of "guesses" to come closer to doing a global optimization (avoid getting stuck in local minima)

After this, the 398 recordings I tested (each between 5 and 20 minutes) all match MaxFilter very well. In places that they don't, our code produces a less "noisy" result, which is an improvement as I have seen MaxFilter's estimates oscillate between two position estimates at times.

@bloyl there are a ton of changes and fixes here, it would be good if you could re-check your use cases to see if they still work properly. I did have to change some of the "spot check" values in the tests. Most were subtle, but one changed from `[0.054, 0.009, 0.037]` with GOF `0.075` to `[-0.065, 0.075, 0.000]` with GOF 0.9323 -- so way better GOF but quite far from the old value. I'm inclined to assume that the new value is correct and the old one was wrong.

Also contains changes to ProgressBar that motivated #7155. In this PR things are nicer but not as nice (or maintainable) as in #7155. So my preference is to roll back the changes here and merge #7155, but if we're going to kill #7155 these changes are useful.

### CODE REVIEW

**Comment by larsoner** [comment]
Hit a UTF8 encoding problem with the modernized ProgressBar that #7155 would allow us to avoid having to fix ourselves...

**Comment by larsoner** [comment]
New tutorial showing the API: https://17855-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/preprocessing/plot_59_head_positions.html

File: doc/python_reference.rst [hunk 1]
```diff
@@ -425,6 +425,8 @@ EEG referencing:
    :toctree: generated/
 
    calculate_head_pos_ctf
+   calculate_chpi_coil_locs
+   calculate_head_pos_chpi_coil_locs
```
Comment (agramfort) [inline doc/python_reference.rst]: calculate_head_pos_chpi_coil_locs -> calculate_head_pos or calculate_head_pos_fif  to be consistent with calculate_head_pos_ctf ?

note that we have public function called plot_head_positions

or maybe we could have a single function calculate_head_pos that takes CTF or fif files as input?

naive question: why do you do it in 2 steps for fif and 1 step for CTF?

File: mne/fixes.py [hunk 2]
```diff
@@ -1250,6 +1252,25 @@ def bincount(x, weights, minlength):  # noqa: D103
...
+                result[i] = func1d(arr[:, i])
+        else:
+            result = np.empty(arr.shape[0])
+            for i in range(len(result)):
+                result[i] = func1d(arr[i, :])
+        return result
+
+    @jit()
+    def mean(array, axis):
+        return _np_apply_along_axis(np.mean, axis, array)
```
Comment (agramfort) [inline mne/fixes.py]: is it really faster then doing np.mean with axis directly?

Comment (larsoner) [reply mne/fixes.py]: No it's the same speed. But this is a necessary shim to be able to do `np.mean(..., axis=...)` *inside* Numba `jit`-ed functions. And `jit`ting those functions (which require `np.mean(..., axis=...)` support) is substantially faster than staying in pure Python.

Someday when Numba adds support for the `axis` argument, we can remove this shim

**Comment by agramfort** [comment]
ok !

File: doc/python_reference.rst (hunk 1 above)
Comment (larsoner) [reply doc/python_reference.rst]: > note that we have public function called plot_head_positions

We have `read_head_pos`, `write_head_pos`, and `head_pos` is the arg for `maxwell_filter` so really `plot_head_positions` is the outlier

> calculate_head_pos_chpi_coil_locs -> calculate_head_pos or calculate_head_pos_fif to be consistent with calculate_head_pos_ctf ?
>
> or maybe we could have a single function calculate_head_pos that takes CTF or fif files as input?
> 
> naive question: why do you do it in 2 steps for fif and 1 step for CTF?

cHPI head position estimation is a two-step process:

1. calculating the locations of the N HPI coils as a function of time, and then
2. using these to calculate time-varying dev_head_t

`calculate_head_pos_ctf` only does the second step, because CTF has already done the first one, namely computed the coil locations, and stored the results in their file. When you call `calculate_chpi_coil_locs` (which can be done for FIF, Artemis123, or any system that has coils continuously emitting HPI freqs -- maybe even CTF?), you are doing the process that CTF has already done.

The naming is meant to reflect this: `calculate_chpi_coil_locs` does the first step, and either `calculate_head_pos_ctf` or `calculate_head_pos_chpi_coil_locs` will compute a `head_pos` depending on whether you want to use the CTF coil location estimates or our own `chpi_coil_locs`, respectively.

As to why it should be separated into two steps instead of a single one:

1. The first step is *much* slower than the second step. It can take 5 minutes to compute coil locations for a 20 minute recording with a lot of movement.
2. Just having the coil positions themselves (output of the first step) can be useful for QA, as you can get a quick time-varying cHPI quality assessment based on the inter-coil distannces (we do this at UW). And the `coil_locs` output is just a dict of `ndarray` so it can easily be written to disk and read back with with `h5io`.
3. The first step in general you won't change the parameters. The defaults are good and reasonable, so you should only really need to run it once per recording.
4. The second step is much faster (5-10 sec for a 20 minute recording) but in general is the one where you might need to iteratively / interactively tweak params to get good results -- `adjust_dig`, `gof_limit`, and `dist_limit` all matter, especially for child recordings where they move a lot.

Comment (agramfort) [reply doc/python_reference.rst]: I think it's better if we have same the functions that deal internally with the fact that it's ctf or fif, eventually with optional coil locs if already present in the file. So

mne.chpi.compute_chpi_coil_locs or maybe mne.chpi.compute_chpi_locs ?
mne.chpi.comput_head_pos that can take ctf or fif and requires chpi_locs as parameters if fif is passed.

and then head_pos can be passed to maxwell_filter.

note that to be consistent it should be called compute and not calculate.

wdyt?

Comment (larsoner) [reply doc/python_reference.rst]: I don't like the distinction for FIF vs CTF. It's not about FIF/Elekta because the coil locs can be for FIF/Elekta, Artemis123, or even recomputed in theory for CTF (and also new systems if they ever come out). I'd rather abandon the vendor specific naming and stay with what they actually functionally do:

- compute_chpi_locs
- compute_head_pos_chpi_locs
- compute_head_pos_ctf_chpi_locs

Or so. Or maybe make it so that there is a `extract_ctf_coil_locs` and then for CTF data you call this and then call `compute_head_pos_chpi_locs`, i.e.:

- compute_chpi_locs
- extract_chpi_locs_ctf
- compute_head_pos_chpi_locs

Comment (larsoner) [reply doc/python_reference.rst]: Actually the last proposal seems cleanest, and this naming is even better:

- compute_chpi_locs (for any vendor)
- extract_chpi_locs_ctf (only for ctf)
- compute_head_pos(coil_locs, ...) (for any vendor)

Comment (drammock) [reply doc/python_reference.rst]: This last one is the API I like best, out of the ones discussed here so far.

Comment (larsoner) [reply doc/python_reference.rst]: Looking at the code it also makes things more DRY as there was duplicated (and outdated) code in `calculate_head_pos_ctf` that now lives in a unified `compute_head_pos` function.

Comment (larsoner) [reply doc/python_reference.rst]: Pushed a commit with this API, @agramfort feel free to see if you agree it's cleaner

**Comment by agramfort** [comment]
i like the new API solution !

to check the quality of the results vs MF can you report a plot_head_positions using either MF or mne-python on 2-3 subjects?

thanks heaps !

out of curiosity is the python code faster than the maxfilter command?

**Comment by larsoner** [comment]
Here is the worst match I got across the 398 files. For this recording (11 min of noisy infant data with a lot of movement), head position estimation takes MaxFilter:
```
real	8m16.905s
user	59m18.603s
sys	0m25.590s
```
and in Python:
```
real	6m14.985s
user	6m7.895s
sys	0m1.311s
```
Plots for Python:

Screenshot from 2020-02-06 15-23-20

Then MaxFilter:

Screenshot from 2020-02-06 15-23-24

I think Python is probably doing a better job on this one of not oscillating between estimates for example.

Adult data looks great:

Screenshot from 2020-02-06 15-27-07

Screenshot from 2020-02-06 15-27-05

**Comment by agramfort** [comment]
@bloyl or @wmvanvliet maybe you want to have a look / try before we merge?

**Comment by larsoner** [comment]
@bloyl should definitely look before merge since some values have changed (@bloyl see top comment for explanation).

I also pushed a commit to remove the changes to `progressbar.py` (saved [here](https://gist.github.com/larsoner/04bb719dc450327d89fa192f7800b452)) so we don't have to wait for a decision on #7155 to merge this one (and also because hopefully we won't need my changes because `tqdm` has more features, probably does it better, etc.).

**Comment by wmvanvliet** [comment]
Tutorial looks awesome! I'll try on some of my own data.

**Comment by wmvanvliet** [comment]
I'm getting a difference in x-offset between maxfilter and MNE-Python.

maxfilter:
maxwell_pos

MNE-Python:
mne_pos

**Comment by larsoner** [comment]
That offset appears to be 1 mm. I find that acceptable.

**Comment by wmvanvliet** [comment]
fair enough

**Comment by bloyl** [comment]
Api changes look good. I'll hopefully have time today to look closer and to
test on some Artemis data.

**Comment by larsoner** [comment]
Great, thanks for looking and testing @wmvanvliet and @bloyl

**Comment by bloyl** [comment]
On artemis data I don't have ground truth to compare with so quantitative testing is harder.

Qualitatively, i've looked and computing dev_head transforms a few of our datasets and I'd say that when I have good data the fits are qualitatively similar.

One issue I did have is when I have bad data the new code returns a significantly higher GOF which makes it hard to catch the bad data and the suspect `dev_head` transforms.

For instance in the following figure shows the topo for the 4 coils. the 1st and 4th have good topos and localize, unfortunately the 2nd and 3rd don't give good localizations but still have `GOFs > 0.99`. In the old code their GOFs were more like `0.93` and were easily excluded.

artemis_example_1

I don't know if this is a problem or an improvement of the GOF code but it is a difference.

@larsoner if your interested I can probably share a dataset and code with you.

Also I think i can acquire a CTF dataset that would enable a comparison between CTF fitting code and the MNE code. I hope to get on the scanner early next week to acquire it.

**Comment by larsoner** [comment]
My guess is that those coils are actually being localized (orientally much) better than they were before (not getting stuck in local minima). The locations might actually be usable now.

If they're not, you might consider quantifying the coil SNR (can use PSD at the court frequencies we signal and midlines between them as noise for example), or a pairwise distance metric instead. The payee distances would give you a quick "how good are these" relative to ground truth, assuming the could have not moved on the head the osteosarcoma distances should stay fixed.

**Comment by larsoner** [comment]
But yes feel free to share code and I can do some coil quality quantification.

Along the lines of annotate_movement, I've been using pairwise coil distance code to estimate the number of usable coils and could make a PR to add it at some point. We also quantify coil SNR as a function of time and I could add that, too.

**Comment by larsoner** [comment]
> I don't know if this is a problem or an improvement of the GOF code but it is a difference.

Back to a keyboard where I can actually type. In other words:

1. I suspect that it's likely an improvement in the fitting optimization that it found a location that explains 99%+ of the variance (instead of 93% before; probably from point 7 above). I didn't change the GOF calculations themselves in this PR. You can disable the "guesses" in the code pretty easily by modifying the conditional to be something like `if guesses is not None and False:`, then it won't do the global thing.
2. You could try running with `ext_order=0` to see if it changes things (point 6 above) but it probably won't make too much difference. It's possible that you had some homogeneous fields that were getting you only 93% explained before, and those are removed now, and using `ext_order=0` should undo this behavior.
3. Is it possible that the coils actually were rotated down below the helmet to the back, and previously they were being localized incorrectly but now they are correctly localized (or at least localized better than before)? 

The pairwise distances might give us some insight, so if you can share that potentially coil-2-/coil-3-problematic dataset that would help. We should see that in `master` and this PR that the pairwise distances for coils 1 and 4 are good in both cases, and I suspect that we'll see that they are bad for coils 2 and 3 (at least to coils 1 and 4) in `master` but better in this PR.

I'm happy to check these things if you share the data.

**Comment by larsoner** [comment]
From talking to @bloyl it seems likely that having access to the actual amplitude estimates could also be useful for QA purposes. So the steps are now split just a little bit more:

- `chpi_amplitudes = compute_chpi_amplitudes(raw)` (for any vendor)
- `chpi_locs = compute_chpi_locs(raw.info, chpi_amplitudes)` (for any vendor)
- `chpi_locs = extract_chpi_locs_ctf(raw)` (only for ctf)
- `head_pos = compute_head_pos(chpi_locs)` (for any vendor, from either of the two `chpi_locs` returning functions above)

Making it so that `chpi_amplitudes` is returned will eventually make it easier to tell for example if one coil has much lower amplitudes than the others, suggesting that it might be much farther from the helmet.

@bloyl in changing this API I found ways of unifying things again, can you look in particular at `artemis123.py` to see if you're okay with the changes there? Basically we fake a `hpi_result` to make `_get_initial_hpi_fit` happy in a way that should be backward compatible with what you were doing before. This has the advantage of getting rid of the need for a separate `_fit_device_hpi_positions` function (now we can just use `compute_chpi_amplitudes` and `compute_chpi_locs` directly).

I also added some basic pairwise-distance code back in, as I think eventually @bloyl might want it for Artemis123, we'll probably want it for QA code, and it's another nice sanity check.

**Comment by wmvanvliet** [comment]
From an regular user point of view: I'm fine with the extra API steps as long as the `maxwell_filter` function can do continuous head position correction "automatically", meaning I can just set a flag to `True` and it will do it, just like the `maxfilter` command line program.

**Comment by larsoner** [comment]
> From an regular user point of view: I'm fine with the extra API steps as long as the maxwell_filter function can do continuous head position correction "automatically", meaning I can just set a flag to True and it will do it, just like the maxfilter command line program.

I'd rather not do this for the following reasons:

1. We are already headed in a separation-of-functionality direction with other aspects of MF. For just two examples (there might be more):
    - `filter_chpi` -- you have to call this separately before doing `maxwell_filter` to get what MaxFilter will do automatically and by default for you.
    - MaxFilter always writes files but we do not have `out=True` by default (or even an option), so users are already doing `raw_sss.save` manually afterward.
2. `head_pos` in `master` is currently a separate step because we don't support, so I think the barrier for adoption / annoyance level with still needing to make separate call(s) should stay low (at least now they're in Python!).
3. Adding `head_pos=True` by itself will leave implicit a bunch of choices that you cannot change, or if we add the `kwargs` to change them to `maxwell_filter`, it will further blow up the already long API for `maxwell_filter`.
4. If people want a single monolithic equivalent, we can later add a new `mne maxwell_filter`, possibly mimicking the MaxFilter API.

TL;DR: I'd argue in the opposite direction, that `maxwell_filter` is already too monolithic and ideally would have been broken down into separate steps from the start (but actually a number of the steps are not easily separable, such as the destination head position until we add a `FIFFV_MULTIPOLE_MOMENT` coil type, tSSS from time-varying SSS decomposition, eventually #7156, etc.).

File: doc/changes/latest.inc [hunk 3]
```diff
@@ -19,7 +19,9 @@ Changelog
 
 - Add current source density :func:`mne.preprocessing.compute_current_source_density` to compute the surface Laplacian in order to reduce volume conduction in data by `Alex Rockhill`_
 
-- Add :func:`mne.chpi.calculate_head_pos_ctf` by `Luke Bloy`_
+- Add :func:`mne.chpi.extract_chpi_locs_ctf` to extract cHPI coil locations from CTF data  by `Luke Bloy`_
+
+- Add :func:`mne.chpi.compute_chpi_locs` and :func:`mne.chpi.compute_head_pos` to compute head positions from cHPI coil locations by `Eric Larson`_ and `Luke Bloy`_
```
Comment (agramfort) [inline doc/changes/latest.inc]: compute_chpi_amplitudes is missing

Comment (larsoner) [reply doc/changes/latest.inc]: Good catch, fixed

File: mne/chpi.py [hunk 4]
```diff
@@ -603,284 +559,156 @@ def _fit_cHPI_amplitudes(raw, time_sl, hpi, fit_time, verbose=None):
         # the first component holds the predominant phase direction
         # (so ignore the second, effectively doing s[1] = 0):
         sin_fit[fi, :] = vt[0]
```
Comment (bloyl) [inline mne/chpi.py]: This seems to return unit strength amplitudes for each coil and each time point. would it be possible to maintain that information. That would let you look at relative hpi signal strength between coils and over time. It would also make the amplitude of the magnetic moment that we end up fitting meaningful. 

Without this `compute_chpi_amplitudes` is much less useful for QA

File: mne/chpi.py [hunk 5]
```diff
@@ -931,126 +783,221 @@ def _calculate_chpi_coil_locs(raw, t_step_min=0.1, t_step_max=10.,
...
+    del fwd, R
+
+    iter_ = list(zip(sin_fits['times'], sin_fits['slopes']))
+    chpi_locs = dict(times=[], rrs=[], gofs=[])
+    # setup last iteration structure
+    hpi_dig_dev_rrs = apply_trans(
+        invert_transform(info['dev_head_t'])['trans'],
+        _get_hpi_initial_fit(info, adjust=adjust_dig))
+    last = dict(sin_fit=None, coil_fit_time=sin_fits['times'][0] - 1,
+                coil_dev_rrs=hpi_dig_dev_rrs)
```
[the remark this answers is not in the document]
Comment (larsoner) [reply mne/chpi.py]: With the `guesses` code in there now using a grid, it probably does not matter much unless you are quite far out of the helmet (and even then might not matter)

File: mne/chpi.py (hunk 4 above)
Comment (larsoner) [reply mne/chpi.py]: Commit pushed to multiply `s[0]` and ensure that the resulting `slopes` are at least somewhat reasonable

File: mne/tests/test_chpi.py [hunk 6]
```diff
@@ -210,53 +241,96 @@ def test_calculate_chpi_positions():
...
+    proj, _, _ = _setup_ext_proj(raw.info, ext_order=1)
+    chpi_amplitudes = dict(times=np.zeros(1), slopes=slopes, proj=proj)
+    chpi_locs = compute_chpi_locs(raw.info, chpi_amplitudes)
+
+    # check GOF
+    coil_gof = raw.info['hpi_results'][0]['goodness']
+    assert_allclose(chpi_locs['gofs'][0], coil_gof, atol=0.3)  # XXX not good
+
+    # check moment
+    # XXX our forward and theirs differ by an extra mult by _MAG_FACTOR
```
Comment (larsoner) [inline mne/tests/test_chpi.py]: @bloyl coil moments added. Feel free to look and merge if you're happy

As a side note, I don't know why our calculation and MaxFilter's differ by `_MAG_FACTOR` (1e-7). I'm guessing that they accidentally double-multiply by it? cc @agramfort in case you know.

**Comment by bloyl** [comment]
Thanks @larsoner
